contracts/contracts/coordination/InfractionCollector.sol

```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import "@openzeppelin-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import "./Coordinator.sol";
+import "../../threshold/ITACoChildApplication.sol";
+
+contract InfractionCollector is OwnableUpgradeable {
+    Coordinator public coordinator;
+
+    // Reference to the TACoChildApplication contract
+    ITACoChildApplication public tacoChildApplication;
+
+    // infraction types
+    enum InfractionType {
+        MISSING_TRANSCRIPT
+    }
+
+    // Mapping to keep track of reported infractions
+    mapping(uint32 => mapping(address => mapping(InfractionType => bool))) public infractions;
+
+    function initialize(
+        Coordinator _coordinator,
+        ITACoChildApplication _tacoChildApplication
+    ) public initializer {
+        __Ownable_init(msg.sender);
+        coordinator = _coordinator;
+        tacoChildApplication = _tacoChildApplication;
+    }
+
+    function reportMissingTranscript(
+        uint32 ritualId,
+        address[] calldata stakingProviders
+    ) external {
+        // Ritual must have failed
+        require(
+            coordinator.getRitualState(ritualId) == Coordinator.RitualState.DKG_TIMEOUT,
+            "Ritual must have failed"
+        );
+
+        for (uint256 i = 0; i < stakingProviders.length; i++) {
+            // Check if the infraction has already been reported
+            require(
+                !infractions[ritualId][stakingProviders[i]][InfractionType.MISSING_TRANSCRIPT],
+                "Infraction already reported"
+            );
+            Coordinator.Participant memory participant = coordinator.getParticipantFromProvider(
+                ritualId,
+                stakingProviders[i]
+            );
+            if (participant.transcript.length == 0) {
+                // Transcript TX wasn't posted
+                // Penalize the staking provider
+                tacoChildApplication.penalize(stakingProviders[i]);
+                infractions[ritualId][stakingProviders[i]][
+                    InfractionType.MISSING_TRANSCRIPT
+                ] = true;
```

Comment:
Related to my above comment - it's worth considering an alternate structure here so that it's easier to lookup all infractions pertaining to a single node.  Without any real world usage data, it's hard to know exactly what types of on-chain queries we will need to handle penalties later.  Based on prior discussions, we may need to identify multiple subsequent infractions in order to fairly enforce a downtime penalty.

Reply:
taco app is checking subsequent penalties, so we can put it there, otherwise structure will change to mapping and arrays and will consume more gas (not a end of the world though)